promql/engine.go

```diff
@@ -1918,27 +1920,46 @@ var (
 	hPointPool zeropool.Pool[[]HPoint]
 )
 
-func getFPointSlice(sz int) []FPoint {
+// pointsSliceSize calculates a reasonable initial capacity for the point slices when running a query
+// taking in consideration the remaining number of steps, maxSamples and number of series in the result
+func (ev *evaluator) pointsSliceSize(numSteps, step, numberOfSeries int) int {
+	// Subtract the current step from the numSteps as at this point we know the given series does not have data before this step
+	remainingSteps := numSteps - step
+	// spread the remaining allowed samples across all the series in the result.
+	allowedSamples := (ev.maxSamples - ev.currentSamples) / numberOfSeries
```

Comment:
I am not sure this complex heuristic is really needed. It guesses a lot, plus it requires number of series, which blocks streaming approaches at some point (nit). I wonder if simple `return remainingStep if it's smaller than (ev.maxSamples - ev.currentSamples)` is not enough for a start?

Reply:
ok ok... this also make sense!

Reply:
oh.. would this work?

The problem here is it will cause the same problem.. cause the current samples will be very low for sparse series, and we will end up allocating the whole steps to all the series.

Reply:
Got it, sorry misunderstood the root cause within that suggestion. We can have trillion steps, but if we have data only for subset of those, we should still return query just fine.

Still, in our main discussion we mentioned the logic of `min(5K, step)` pre-alloc (or 10k). This could work fine for start. If we are worried about that very sparse case, we add recompact slice logic after we fill it with points if really needed (although question is how long this slice will live). Your idea with `min(((ev.maxSamples - ev.currentSamples) / numberOfSeries), step)` is not bad, just it's pure guess and can preallocate a lot still. It's even a bit weird that the more series we process the less we preallocate? Anyway, both fine I guess, just the simpler is equally efficient in generic unknown environments?

Reply:
> It's even a bit weird that the more series we process the less we preallocate

Yeah... but this is the downside but we have to keep in mind that if we are running far from the `ev.maxSamples` this condition will never happen... And if we are running close to it we are allocating less per series and so give change to hit that limit before going OOM.

I think to prevent what we saw we could do `min(5K, step)` or we need to divide by the number of series.

One thing i wanna point out is that queries over sparse series in this case are not uncommon if we are talking about 30 days worth of data (if you deploy once every 2 days you will have 2 days worth of data on mainly of your series). And for those series we are always allocating slices to fit 30 days steps even though we only have data for 2 days.